sdk-rs/src/dlob.rs

```diff
@@ -110,6 +116,78 @@ impl DLOBClient {
             }
         });
 
+        RxStream(rx)
+    }
+
+    /// Subscribe to an orderbook via WebSocket.
+    pub async fn subscribe_ws(&self, market: MarketId) -> L2OrderbookStream {
+        // This unwrap should never panic
+        let ws_url = crate::utils::http_to_ws(&self.url).unwrap();
+        let (mut ws_stream, _) = connect_async(ws_url).await.unwrap();
+
+        // Setup channel for L2OrderbookStream
+        let (tx, rx) = channel::<Result<L2Orderbook, SdkError>>(16);
+
+        match market.kind {
+            MarketType::Perp => {
+                if let Some(perp_market_config) = perp_market_config_by_index(self.context, market.index) {
+                    let market_subscription_message = to_ws_json(perp_market_config);
+                    ws_stream.send(Message::Text(market_subscription_message)).await.map_err(|_| SdkError::SubscriptionFailure).unwrap();
+                }
+            },
+            MarketType::Spot => {
+                if let Some(spot_market_config) = spot_market_config_by_index(self.context, market.index) {
+                    let market_subscription_message = to_ws_json(spot_market_config);
+                    ws_stream.send(Message::Text(market_subscription_message)).await.map_err(|_| SdkError::SubscriptionFailure).unwrap();
+                }
+            }
+        }
+
+        let heartbeat_interval = TokioDuration::from_secs(5);
+        let mut last_heartbeat = Instant::now();
+        tokio::spawn(async move {
+            while let Some(message) = ws_stream.next().await {
+                if last_heartbeat.elapsed() > heartbeat_interval {
+                    eprintln!("Heartbeat missed!");
+                    let _ = ws_stream.close(None).await;
+                    break;              
+                }
+
+                match message {
+                    Ok(Message::Text(text)) => {
+                        let value: Value = serde_json::from_str(&text).unwrap_or_else(|_| json!({}));
```

Comment:
recommend this crate which is better at extracting small parts of json in this style https://github.com/tidwall/gjson.rs

otherwise I think serde_json::from_str tries to parse the whole payload so we should aim to use strongly typed payloads for better performance

Reply:
isn't the payload that's being sent like 95% required data ? printed, it's bids, asks, slot, and then a small amount of market info at the very end

Reply:
ok let's leave it, I was expecting `data` to be a JSON object but since its sent as a string this is fine